### PR TITLE
#940 - Typings for component has been added

### LINKED ISSUE #940: Missing typings for Component class
**Component** just extends **HTMLElement** but has no methods (callbacks) declared.

### PR DESCRIPTION
**Pull request by hopkins-tk**
Callbacks added to **Component** type definition.

Fixes #940

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -6,7 +6,14 @@ export function props(elem: any, data?: PropData): void;
 
 export const emit: (elem: any, eventName: string, eventOptions = {}) => boolean;
 
-export class Component extends HTMLElement {}
+export class Component extends HTMLElement {
+  updatedCallback?(prevProps: { [name: string]: any }): boolean;
```
Comment (Hotell) [inline src/index.d.ts]: why do we need those methods as optional?

Comment (andreawyss) [inline src/index.d.ts]: Because the implementer of the Component is not required to provide one.

Comment (Hotell) [inline src/index.d.ts]: that is not true @andreawyss, if it was interface and user would use `implements`, then it would errored, not with this one.

By having it on class definition, user is not forced to use that, it just provides better intellisense  within editor.

File: src/index.d.ts [hunk 2]
```diff
@@ -6,7 +6,14 @@ export function props(elem: any, data?: PropData): void;
...
+  renderCallback?(): void;
```
Comment (Hotell) [inline src/index.d.ts]: this is wrong definition. `renderCallback` has to return `null` or `JSX.Element` or `vdom`

File: src/index.d.ts (hunk 1 above)
Comment (andreawyss) [inline src/index.d.ts]: @Hotell Yes, you are correct.

File: src/index.d.ts (hunk 2 above)
Comment (hopkins-tk) [inline src/index.d.ts]: I had glanced to the tests and there are even arrays being returned for this callback, so I returned the generic `any | undefined` that is used elsewhere as well.

So I am not sure what are all the possibilities.

File: src/index.d.ts [hunk 3]
```diff
@@ -6,7 +6,14 @@ export function props(elem: any, data?: PropData): void;
 
 export const emit: (elem: any, eventName: string, eventOptions = {}) => boolean;
 
-export class Component extends HTMLElement {}
+export class Component extends HTMLElement {
```
Comment (Hotell) [inline src/index.d.ts]: I would add whole API although it doesn't affect consumption, just to be consistent with implementation, until skate is migrated to TS 
```ts
class Component extends HTMLElement {
  static readonly props: { [name: string]: PropOptions };
  static readonly observedAttributes: string[];
// ....
}
```

File: src/index.d.ts [hunk 4]
```diff
@@ -6,7 +6,14 @@ export function props(elem: any, data?: PropData): void;
...
+  updatedCallback(prevProps: { [name: string]: any }): boolean;
+  renderCallback(): any | undefined;
+  renderedCallback(): void;
+  connectedCallback(): void;
+  disconnectedCallback(): void;
```
Comment (andreawyss) [inline src/index.d.ts]: updatedCallback(prevProps: { [name: string | symbol]: any }): boolean;

renderCallback(): any | any[] | undefined;

Comment (Hotell) [inline src/index.d.ts]: there absolutely no reason to do `any | any[]`, because it merges just to `any`.

Those definitions needs to be rewritten with appropriate `JSX` namespace but that's not part of this PR

Comment (Hotell) [inline src/index.d.ts]: also `renderCallback` shouldn't ever return `undefined` , the same as in preact or react. If it doest it's a programatic error

**Comment by andreawyss** [comment]
Everywhere a prop name is passed should be defined as:  string | symbol

**Comment by treshugart** [comment]
Can you change the destination branch to the "typescript" branch, please? Unfortunately GitHub doesn't allow you to do this, so you'll have to resubmit the PR. Sorry :(

**Comment by Hotell** [comment]
> Everywhere a prop name is passed should be defined as: string | symbol

this isn't possible yet @andreawyss 
https://github.com/Microsoft/TypeScript/issues/1863

**Comment by Hotell** [comment]
I would merge this as is, and continue with further type definitions tweaking, possibly with @vvakame contributions. wdyt? @treshugart @andreawyss

**Comment by andreawyss** [comment]
I would just do this more:

attributeChangedCallback(name: string, oldValue: **null|string**, newValue: **null|string**): void;

static attributeChanged?(elem: Component, data: { name: string, oldValue: **null|string**, newValue: **null|string** }): void;

and then this looks good to me.

**Comment by hopkins-tk** [comment]
I have changed the values types to **string** - I don't think **null** is needed, as all the fields are _nullable_.

Or do you feel that explicit **null** would be any good there?

**Comment by Hotell** [comment]
if user uses `strictNullTypeChecks` we need to explicitly annotate types which should be nullable with `null` type, instead he would get error, that's why we need to be explicit

**Comment by andreawyss** [comment]
I will merge this into typescript branch and then fix there before going top master
